### DISCUSSION TITLE
CRC errors and no DEBUG with power meter and RTU example code

### DISCUSSION THREAD

**Discussion by brainstorm** [eModbus/eModbus#166] [Q&A] [answered]
Hello, I'm new here, coming from https://github.com/eModbus/eModbus/issues/136#issuecomment-1068041613 (see minimodbus code and output in unfoldable details there)...

I'm now struggling with some CRC errors and after reading through this forum I think I need to:

1) Adjust the bus biasing properly, following the [great](https://github.com/eModbus/eModbus/discussions/112) [discussions](https://github.com/eModbus/eModbus/discussions/110)... adapted to my MAX481 instead of the MAX485.
2) Enable `-DLOG_LEVEL=LOG_LEVEL_VERBOSE` in the `platform.ini` common build_flags to overcome the CRC errors I'm having:

```
[E] 1635| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[E] 2636| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[E] 3637| main.cpp             [  59] handleError: Error response: E2 - CRC check error
(...)
```

I can't seem to have an HexDump of the packets on the bus as shown in other discussions, what am I doing wrong? From what I'm reading in the library code, the error handler should have hexdump macros too...

Unfortunately I'm on holidays and don't have an oscilloscope/LA at reach to see what's going on electrically :-S Here's my relatively precarious setup with a couple of resistors, one of them biasing to GND with 220Ohm:

![IMG_2076](https://user-images.githubusercontent.com/175587/158606577-7c67067d-d53d-4ea1-b136-9832788c66c9.JPG)
![IMG_2077](https://user-images.githubusercontent.com/175587/158606494-5bae651e-d76b-4120-852e-7ade75fa84ac.JPG)

**Comment by Miq1** [comment]
Please try again with ``-DLOG_LEVEL=6`` in the ``build_flags`` of ``platformio.ini``. It may be that the ``#define LOG_LEVEL_VERBOSE 6`` is not yet read at that point.
If the verbose log is successfully activated, you should see a "Raw buffer" line in it with the received bytes.

**Comment by brainstorm** [comment]
Thanks, progress, see commit https://github.com/brainstorm/yigedianqi-modbus/commit/c8a16b8cca12a995ce4af08243526d85908f70e8! Now I see proper hexdumps :D

```
[E] 1636| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[V] 2614| ModbusMessage.cpp    [ 529] checkData: Check data #3
[D] 2614| ModbusClientRTU.cpp  [ 126] addRequestM: request for 01/04
[D] 2615| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 2615| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 2616| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FF9F7A0/6:
  | 0000: 01 04 00 01 00 08                                 |......          |
[D] 2627| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFC348C/7:
  | 0000: 7F DF FD FF EF BF DE                              |......         |
[D] Received packet: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 2633| ModbusClientRTU.cpp  [ 204] handleConnection: Error response (1 bytes) received.
[V] Data: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 2636| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
[V] Response packet: @3FF9F7A0/3:
  | 0000: 01 84 E2                                          |...             |
```

**Reply by Miq1** [reply]
Tha data is looking rather odd, like it is the 2's complement of the real data or something.
```
7F DF FD FF EF BF DE
```
is no valid Modbus packet. The length does not match as well, you were requesting 8 registers, so you should have received 8*2+3=19 bytes instead of 7 only. Do you have the baud rates set to the same value on client and server?

**Reply by brainstorm** [reply]
Agreed, those values are a bit odd... The BAUDRATE seems to match with the physical device config and the sketch, 9600, and the ModBus addr is 1, see:

![IMG_2079](https://user-images.githubusercontent.com/175587/158632706-b4e3df05-091a-47bb-86df-e3a5a9c1a637.JPG)
![IMG_2080](https://user-images.githubusercontent.com/175587/158632794-863f116e-dd6a-4217-8b8c-d5c201eefc37.JPG)

But at least those values are consistent across re-flashing/runs, so there's no noise or other arbitrary confounders are involved?

Then, when I change the first register, the returned value seems to vary slightly, for instance:

```diff
diff --git a/src/main.cpp b/src/main.cpp
index 5ed0acc..e4d08c2 100644
--- a/src/main.cpp
+++ b/src/main.cpp
@@ -21,7 +21,7 @@
 #define REDEPIN GPIO_NUM_7
 #define BAUDRATE 9600
 //#define FIRST_REGISTER 0x002A
-#define FIRST_REGISTER 0x0001
+#define FIRST_REGISTER 0x001d
 #define NUM_VALUES 4
 #define READ_INTERVAL 1000
```

Yields:

```
[E] 4667| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[V] 5663| ModbusMessage.cpp    [ 529] checkData: Check data #3
[D] 5663| ModbusClientRTU.cpp  [ 126] addRequestM: request for 01/04
[D] 5664| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FFC348C/6:
  | 0000: 01 04 00 1D 00 08                                 |......          |
[D] 5673| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] 5664| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 5674| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] Raw buffer received: @3FFC348C/7:
  | 0000: 7F DF C5 FF EF 3D 6B                              |....=k         |
[D] Received packet: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 5680| ModbusClientRTU.cpp  [ 204] handleConnection: Error response (1 bytes) received.
[V] Data: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 5682| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
[V] Response packet: @3FF9F7A0/3:
  | 0000: 01 84 E2                                          |...             |
```

So the received packet changed from `7F DF FD FF EF BF DE` to `7F DF C5 FF EF 3D 6B`... I'll change things a bit to see if there's a pattern, but feel free to suggest tests, I'm excited :)

**Reply by brainstorm** [reply]
IIRC, this device was holding values in pairs of registers, see my minimodbus code:

```python
        for reg in range(23,80,2):
            val = instrument.read_float(reg, functioncode=3, numberOfRegisters=2)
            print("Float register {}, {} has value: {}".format(reg, reg+1, val))
```

And its outputs:

```
$ ./minimodbus.py 
Float register 23 has value: 0.0
Float register 25 has value: 0.0
Float register 27 has value: 0.0
Float register 29 has value: 99.9990005493
Float register 31 has value: 0.0
(...)
Float register 63 has value: 0.0
Float register 65 has value: 3547.2199707
Float register 67 has value: 3547.2199707
Float register 69 has value: 3547.2199707
(...)
Float register 79 has value: 1.40129846432e-42
```

Maybe I need to read the registers accordingly. I'll explore the eModBus primitives to see if that can be done easily...

**Reply by Miq1** [reply]
That is not relevant for the current issue you are experiencing. Regardless of data requested, the first three bytes of the response must match server ID, function code and number of payload bytes, so in your example the response must start ``01 04 10 ...``.

If I invert and reverse the bits in your received data, I get 0x7F --> 0x01, 0xDF --> 0x04, 0xFD --> 0x40 ..., which comes amazingly close. There is something wrong with the A/B wiring.

**Reply by brainstorm** [reply]
Makes sense, I'll revisit bus biasing and report back, thanks for the super quick feedback!... oh, just read your last (edited) sentence about inverting and reversing... I'll try also flipping the A/B bus leads then?

**Reply by brainstorm** [reply]
For the record, I physically disconnected the boards (ESP32S2 and RS485 transceiver) from the power meter and the same message keeps on going:

```
[E] 1505731| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[V] 1506722| ModbusMessage.cpp    [ 529] checkData: Check data #3
[D] 1506722| ModbusClientRTU.cpp  [ 126] addRequestM: request for 01/04
[D] 1506723| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 1506723| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 1506724| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FF9F7A0/6:
  | 0000: 01 04 00 1D 00 08                                 |......          |
[D] 1506735| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFC3690/7:
  | 0000: 7F DF C5 FF EF 3D 6B                              |....=k         |
[D] Received packet: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 1506741| ModbusClientRTU.cpp  [ 204] handleConnection: Error response (1 bytes) received.
[V] Data: @3FF9F7A0/1:
  | 0000: E2                                                |.               |
[D] 1506744| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
[V] Response packet: @3FF9F7A0/3:
  | 0000: 01 84 E2                                          |...             |
```

**Reply by Miq1** [reply]
There's definitely noise on your lines. Did you remove the two white patch cables I saw on the picture? These very well may act as antenna!

**Reply by brainstorm** [reply]
I just removed those two white cables you mentioned and the situation improved, when disconnecting it from the device, it goes `E0 - Timeout` and when connecting `E2 - CRC Error`, which makes much more sense than before :)

Perhaps I should just invert and reverse the bits in code now as you hinted? Let's not forget that this device is a dead-brand chinese 60eur thing after all, so I'm not super confident about its adherence to ModBus proper ;)

**Reply by Miq1** [reply]
The bits do not match exactly, so I rather would look after the signal quality first. Another thing could be parity etc. - do you know and match the settings of the meter? 8E1 is the commonest, but may be the meter is different?
On the other hand there seems to be at least a reaction (if the data was not completely fake signals), so the meter then must +ave received something understandable.

**Comment by Miq1** [comment]
There is a very similar device that has a [manual available](https://manualzz.com/doc/6624371/bj194j-9sy-power-analyzer-user-manual-version) that includes Communications and Modbus register specs.

**Reply by brainstorm** [reply]
Very similar indeed! Mine does not have the "bUS" menu where I can select the parity settings though. I know "for sure" that baud on the device is 9600 (as shown on the first screenshots above) but my device does not have the "bUS" menu where things like 8N1 can be selected like the manual states:

<img width="800" alt="Screen Shot 2022-03-16 at 6 19 07 pm" src="https://user-images.githubusercontent.com/175587/158659973-14c4d152-be13-4c81-adb5-1bfd83ce11ab.png">

By the way, I tried them all, from `SERIAL_5E1` all the way to `SERIAL_802`:

<img width="805" alt="Screen Shot 2022-03-16 at 6 21 25 pm" src="https://user-images.githubusercontent.com/175587/158660377-fe2ec048-6692-404d-aa12-875f48a5b438.png">

After building and uploading them one by one, the raw buffer values change but none of them turns 0x7F into 0x01, unfortunately :/

**Reply by Miq1** [reply]
Then the signal path again should be your primary target. There is a RS485 GND connector mentioned in the manual. did you try to connect that to earth potential or even the ESP's GND?
Second, again consider to at least put in a 120 Ohms termination resistor between A and B on the shield's side.
BTW, what voltage is the shield running - 3V3 or 5V?
Last, do you have a scope or a logic analyzer? That may reveal what really is going on on the bus.

**Reply by brainstorm** [reply]
Puzzling side effect... if I set BAUDRATE to 115200 (instead of 9600), the RGB LED on the Saola-1 board turns green :-!

![IMG_2081](https://user-images.githubusercontent.com/175587/158661749-431d584b-b8a9-4f5b-9854-ee6d153d279e.JPG)

I can't explain it by the schematic: https://dl.espressif.com/dl/schematics/ESP32-S2-SAOLA-1_V1.1_schematics.pdf

So I guess it's some kind of register-setting artifact? I'll double check whether I'm using the "wrong" pins that clash somehow with the pin MUX of the Espressif IC.

**Reply by brainstorm** [reply]
> Then the signal path again should be your primary target. There is a RS485 GND connector mentioned in the manual. did you try to connect that to earth potential or even the ESP's GND?

Indeed I did not connect Device GND to Saola's GND, I will focus on that next. It's a bit challenging because there's no GND exposed outside. I'll open up the box and measure the "COM 21" bit (which does not have leads outside):

![IMG_2082](https://user-images.githubusercontent.com/175587/158664473-f39bcab3-1e7b-4da0-b026-3f574e5ff633.JPG)

**Reply by Miq1** [reply]
No, stop that. if there is no official earth connector, it would be a shot into the dark. Strictly speaking RS485 does not need a GND to work, it is advisable (if possible) to earth the device.
Again, at what voltage are you driving the shield?

**Reply by brainstorm** [reply]
Gotcha. The shield is being driven at 3.3V but can be driven up to 5.25V MAX according to MAX481's datasheet.

**Comment by brainstorm** [comment, marked as answer]
There's a [78L05 LDO](https://siliconsupplies.com/media/literature/ss/78L05.pdf) GND, powering a RS485 transceiver, so I could use that for the ESP32S2 board's GND (and perhaps as 5V supply substituting it by a beefier LDO that can do more than 100mA to power up the ESP):

![IMG_2083](https://user-images.githubusercontent.com/175587/158667249-6d66c023-215f-4cf8-81fe-587bdf5ff339.JPG)
![IMG_2086](https://user-images.githubusercontent.com/175587/158667214-de5707a2-9628-48bc-bb59-5a02aa874ffd.JPG)
![IMG_2085](https://user-images.githubusercontent.com/175587/158667350-783dab37-5391-4374-bf22-b4a30cc87b03.JPG)
![IMG_2084](https://user-images.githubusercontent.com/175587/158667295-1d3436d0-8be0-4aed-8fe8-0a5c99af1f89.JPG)

**Reply by Miq1** [reply]
I had a look at the schematics of the shield now: it is definitely run at 5V, whereas the ESP has a 3V3 logic. You will need at least voltage dividers at the TX and RX lines, better a level shifter!

**Reply by brainstorm** [reply]
Thanks for the hint @Miq1!

I've reused a board I built for [JTAG experiments](https://github.com/espressif/openocd-esp32/issues/135#issuecomment-766340269) since it had a proper TTL level converter and rewired it for this use case so that the TX/RX lines have TTL levels between the shield and the ESP, as you suggested:

![IMG_2087](https://user-images.githubusercontent.com/175587/158804401-831619c2-6c5b-4561-822d-8837113e8f42.JPG)
![IMG_2088](https://user-images.githubusercontent.com/175587/158804472-24af5406-00b3-4ca9-a081-f4305b4d3c1d.JPG)

Unfortunately the result is the same: CRC errors :/ I [changed the command from INPUT to HOLD registers](https://github.com/brainstorm/yigedianqi-modbus/commit/d756ff84b2ac99b878129bfadfac39708c0122f0), trying to match [what I had with python](https://github.com/eModbus/eModbus/discussions/166#discussioncomment-2373964), but the non-modbus-compliant response persists:

```
[E] 346342| main.cpp             [  59] handleError: Error response: E2 - CRC check error
[V] 347320| ModbusMessage.cpp    [ 529] checkData: Check data #3
[D] 347320| ModbusClientRTU.cpp  [ 126] addRequestM: request for 01/03
[D] 347321| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 347321| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 347322| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FF9F6F0/6:
  | 0000: 01 03 00 1D 00 02                                 |......          |
[D] 347333| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFC3680/7:
  | 0000: 7F FE C5 FF FB 95 1E                              |......         |
[D] Received packet: @3FF9F6F0/1:
  | 0000: E2                                                |.               |
[D] 347340| ModbusClientRTU.cpp  [ 204] handleConnection: Error response (1 bytes) received.
[V] Data: @3FF9F6F0/1:
  | 0000: E2                                                |.               |
[D] 347342| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
[V] Response packet: @3FF9F6F0/3:
  | 0000: 01 83 E2                                          |...             |
```

I only have a rusty multimeter available, no LA nor oscilloscope, unfortunately (I'm on holidays :). What other signal conditioning should I attempt on this setup? I tried putting a 100Ohm termination resistor on the A/B lines, but there's no change either. I guess that I need to calculate a proper resistor value for biasing the bus as suggested elsewhere... further ideas welcome, I'll continue giving it some thought :)

**Reply by brainstorm** [reply]
Since I'm running out of ideas, I might attempt cross-device (between the power meter and ESP32S2) grounding next, since the manual you shared does mention a 3-wire RS485 setup:

<img width="802" alt="Screen Shot 2022-03-17 at 12 20 56 pm" src="https://user-images.githubusercontent.com/175587/158807315-a0d49950-205a-4505-a084-1e59606210c8.png">

I promise I'll be careful with the hot side and internals ;)

**Reply by Miq1** [reply]
I did warn you...

You might try to put a 4.7k pullup resistor between A and 5V, another as pulldown between B and GND, and an 120 Ohms between A and B. Even if these are not the ideal values they will pull the bus to a stable state.

If the signal is okay and still you are getting these 7-byte messages, please collect a couple and post them here. May be we can make out a pattern.

7 bytes is odd, an error response had 5 (server id, function code, error code and two bytes CRC). A one register response on a read request may fit, but then it would have to be identical for at least the initial three bytes.

**Reply by brainstorm** [reply]
Thanks! Added the resistors as advised:

![IMG_2090](https://user-images.githubusercontent.com/175587/158855807-28d28ff4-f0c9-4f88-889d-0ffc99289361.JPG)

However, I've observed a bad side effect of this arrangement: on the one hand, if I keep all three resistors as advised: 4K7 ohm A+ pullup, 4K7 ohm B- pulldown, 100Ohm termination (didn't have 120ohm handy), the ESP gets stuck in CRC error, **even when the ESP is disconnected altogether from the power meter**.

On the other hand, If I leave the 100Ohm termination and the 4k7 ohm B- pulldown connected and **disconnect** the A+ pullup, I'm getting: **E0 - Timeout when disconnected from the meter** and **E2 - CRC error when connected to the meter** (this latter setup yielding a more logical behavior?).

Perhaps the internal power meter GND is indeed required to stabilize this setup after all? It's definitely challenging to not be able to see the waveforms :/

**Reply by Miq1** [reply]
Rats. This obviously did not help at all. RS485 normally is independent of any GND potential, but just works with the voltage differences between A and B. Provided your shield is working okay the regular 120 Ohms termination should be sufficient and most probably you will find another termination resistor of the same value inside the meter. Only one end of the bus is supposed to be biased against some Vcc and GND potential to maintain a certain voltage difference between A and B.

Too bad you have no scope available - otherwise you could directly look at the A/B voltage difference to see what is happening on the bus.

If the data you are receiving is always 7 bytes, that would speak against electrical bus issues, but hint to some protocol problem. Would you mind collecting a bunch of the received responses and post those here, as I suggested?

**Reply by brainstorm** [reply]
It works now (valid ModBus responses)!!! 

After examining the RS485 circuit from the power meter side, I noticed 10k ohm on the A/B bus and [this electronics post](https://electronics.stackexchange.com/questions/363829/rs422-485-maximum-terminator-resistor-value) was the eye opener I needed to place a 100ohm resistor on the other side, close to the device connector... until reading that post I assumed that there was an internal terminator resistor, but it turned out that 10k is placed there for other reasons (read the linked post for details). 

So, lesson learned: even if the bus cables are barely 5cm long, resistors on each end are crucial to match the conductor's impedance and avoid reflections that confuse the transceivers :)

```
[V] 122980| ModbusMessage.cpp    [ 529] checkData: Check data #3
[D] 122980| ModbusClientRTU.cpp  [ 126] addRequestM: request for 01/03
[D] 122981| ModbusClientRTU.cpp  [ 175] addToQueue: RC=01
[D] 122981| ModbusClientRTU.cpp  [ 137] addRequestM: RC=00
[D] 122981| ModbusClientRTU.cpp  [ 193] handleConnection: Pulled request from queue
[D] Sent packet: @3FF9F7A0/6:
  | 0000: 01 03 00 1D 00 04                                 |......          |
[D] 122992| ModbusClientRTU.cpp  [ 198] handleConnection: Request sent.
[D] Raw buffer received: @3FFC3918/13:
  | 0000: 01 03 08 00 00 00 00 00  00 00 00 95 D7           |.............   |
[D] Received packet: @3FF9F2BC/11:
  | 0000: 01 03 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 123018| ModbusClientRTU.cpp  [ 204] handleConnection: Data response (11 bytes) received.
[V] Data: @3FF9F2BC/11:
  | 0000: 01 03 08 00 00 00 00 00  00 00 00                 |...........     |
[D] 123020| ModbusClientRTU.cpp  [ 225] handleConnection: Response generated.
[V] Response packet: @3FF9F2BC/11:
  | 0000: 01 03 08 00 00 00 00 00  00 00 00                 |...........     |
[V] got float: @3FFC0E88/4:
  | 0000: 00 00 00 00                                       |....            |
[V] got float swapped: @3FFC0E88/4:
  | 0000: 00 00 00 00                                       |....            |
[V] got float: @3FFC0E8C/4:
  | 0000: 00 00 00 00                                       |....            |
[V] got float swapped: @3FFC0E8C/4:
  | 0000: 00 00 00 00                                       |....            |
Requested at  122.980s:
   001D:    0.000
   001F:    0.000
```

So next step would be enumerating the registers and mapping the values from the display correctly, case closed, thanks a ton for your help @Miq1 :)

**Reply by Miq1** [reply]
You are welcome 😉 

Glad to read that you got it running. Come back to ask whenever you like.
